### PR TITLE
ENH:included anti join functionality

### LINKED ISSUE #42916: ENH: Add anti-joins to pandas.merge
#### Is your feature request related to a problem?

Pandas does not allow anti-joins. It would be helpful to have these added
Pyspark has implemented this on its 'join' command.
https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-join.html

#### Describe the solution you'd like

Add anti-left and anti-right joins to 'how' 
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html

#### API breaking implications

None

#### Describe alternatives you've considered

You can replicate an anti-join by doing an outer join, and filtering.

My code does work but sometimes gives too many columns - not sure why. Also my code relies on the column name not already containing '_drop' - which obviously could be an issue.

#### Additional context

[add any other context, code examples, or references to existing implementations about the feature request here]

```python
import pandas as pd
def left_anti_join(df1,df2,left_on,right_on):
   """ Perfoms left anti join of two data frames"""
   df=pd.merge(df1,df2,left_on=left_on,right_on=right_on,suffixes=('','_drop'), how='outer', indicator=True)
   df=df[df['_merge']=='left_only']
   df.drop([col for col in df.columns if '_drop' in col],axis=1, inplace=True)
   df.drop(columns="_merge", inplace=True)
   return df
```

### PR DESCRIPTION
**Pull request by debnathshoham**
- [x] closes #42916
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Extremely naive implementation. Happy to take suggestions.

### CODE REVIEW

File: pandas/tests/reshape/merge/test_merge.py [hunk 1]
```diff
@@ -2487,3 +2487,129 @@ def test_mergeerror_on_left_index_mismatched_dtypes():
     df_2 = DataFrame(data=["X"], columns=["C"], index=[999])
     with pytest.raises(MergeError, match="Can only pass argument"):
         merge(df_1, df_2, on=["C"], left_index=True)
+
+
+@pytest.mark.parametrize(
```
Comment (phofl) [inline pandas/tests/reshape/merge/test_merge.py:2492]: Following things are missing on first glance:

- joining over more than one column
- joining with something like left_index=True, right_on="X"
- merge finds common colls to join over if nothing is provided
- check that nan is matched with nan to be consistent
- what happens with different dtypes?

Comment (attack68) [reply pandas/tests/reshape/merge/test_merge.py:2492]: your tests are very specific and you might miss some of the cases @pholf discusses. You might try multiple `parametrizations` although not too many since you don't want combinations to explode.
e.g.

```python
@pytest.mark.parametrize("left", [{"left_on": "A"}, {"left_index": True}])
@pytest.mark.parametrize("right", [{"right_on": "B"}, {"right_index": True}])
...
kwargs = {**left, **right, **others}
result = merge(**kwargs)
...
```

You may to design a clever routine for expected output based on the different combinations.

Comment (debnathshoham) [reply pandas/tests/reshape/merge/test_merge.py:2492]: Hi @phofl , @attack68 .. I will work on these missing cases. Little difficult to parametrize like that (as the output will be different). Maybe I will setup multiple tests to check specific cases.

Comment (phofl) [reply pandas/tests/reshape/merge/test_merge.py:2492]: Yeah for sure, The outputs vary too much to geht them with one test I think

Comment (debnathshoham) [reply pandas/tests/reshape/merge/test_merge.py]: Hi @phofl @attack68 - I think I have covered all the cases mentioned above. Let me know how this looks.

File: pandas/tests/reshape/merge/test_merge.py [hunk 2]
```diff
@@ -2487,3 +2487,369 @@ def test_mergeerror_on_left_index_mismatched_dtypes():
     df_2 = DataFrame(data=["X"], columns=["C"], index=[999])
     with pytest.raises(MergeError, match="Can only pass argument"):
         merge(df_1, df_2, on=["C"], left_index=True)
+
+
+class Test_AntiJoin:
```
Comment (phofl) [inline pandas/tests/reshape/merge/test_merge.py:2492]: Could you put df1 and df2 into the __init__ method of this class and write down the tests explicitly? These parametrizations are really hard to read. Then you can write down one test for every case without duplicating unnecessary things.

And please call them left and right. Will look more closely after this

Comment (debnathshoham) [reply pandas/tests/reshape/merge/test_merge.py]: I have seperated the tests into another file.. also split the tests for clarity..
Let me know how this looks.

Comment (attack68) [reply pandas/tests/reshape/merge/test_merge.py]: I will have a think on how to improve these tests to make them more compact and readable, but I wonder if issue #43096 might need to be addressed / discussed first, since I think it is relevant for this.

**Comment by attack68** [comment]
I haven't reviewed fully but my instinct is telling me that you have implemented this in a very complicated way, and I think a search/consideration for a much simpler approach is worthwhile. (see my comment in the underlying issue).

Additionally, your solution uses `numpy` and numpy array functions. Does that mean this functionality might fail with `dtypes` that are native to `pandas`?

**Comment by debnathshoham** [comment]
Hi @attack68 , this https://github.com/pandas-dev/pandas/issues/42916#issuecomment-894430300 was really helpful. I have made some changes, let me know how this looks. I believe this should be able to handle pandas dtypes.

File: pandas/core/reshape/merge.py [hunk 3]
```diff
@@ -693,6 +693,9 @@ def __init__(
             self.left_on = self.right_on = [cross_col]
         self._cross = cross_col
 
+        if self.how in ["anti_left", "anti_right", "anti_full"]:
```
Comment (jreback) [inline pandas/core/reshape/merge.py:696]: can you enhance the doc-string of pd.merge.

and add examples

File: pandas/core/reshape/merge.py [hunk 4]
```diff
@@ -744,6 +747,57 @@ def get_result(self) -> DataFrame:
 
         return result.__finalize__(self, method="merge")
 
+    def _anti_join_update(self):
+        def _anti_helper(_left, _right, _how):
```
Comment (jreback) [inline pandas/core/reshape/merge.py:751]: move this helper to its own module level function & fully type in & out arguments & add doc-strings

File: pandas/core/reshape/merge.py [hunk 5]
```diff
@@ -744,6 +747,57 @@ def get_result(self) -> DataFrame:
...
+                    _left = Index(_left.values.flatten())
+                else:
+                    _left = MultiIndex.from_frame(_left)
+            if not isinstance(_right, Index):
+                if len(_right.columns) == 1:
+                    _right = Index(_right.values.flatten())
+                else:
+                    _right = MultiIndex.from_frame(_right)
+
+            if _how in ["anti_left", "anti_right"]:
```
Comment (jreback) [inline pandas/core/reshape/merge.py:763]: too magical, just have if/elif (or use a dict)

File: pandas/core/reshape/merge.py [hunk 6]
```diff
@@ -744,6 +747,57 @@ def get_result(self) -> DataFrame:
...
+                _how = _how.split("_")[1]
+            else:
+                _how = "outer"
+            join_index_l = ~_left.isin(_right)
+            join_index_r = ~_right.isin(_left)
+            return (join_index_l, join_index_r, _how)
+
+        if self.left_index and self.right_index:
```
Comment (jreback) [inline pandas/core/reshape/merge.py:771]: add comments where things are not obvious

File: pandas/core/reshape/merge.py (hunk 3 above)
Comment (jreback) [reply pandas/core/reshape/merge.py:696]: this should be an elif

File: pandas/core/reshape/merge.py [hunk 7]
```diff
@@ -744,6 +747,57 @@ def get_result(self) -> DataFrame:
 
         return result.__finalize__(self, method="merge")
 
+    def _anti_join_update(self):
```
Comment (jreback) [inline pandas/core/reshape/merge.py:750]: ok with this being a method on _MergeOperation but it should have NO side effects, e.g. simply return left, right and assign at this level.

File: pandas/tests/reshape/merge/test_merge_anti.py [hunk 8]
```diff
@@ -0,0 +1,407 @@
+import numpy as np
+import pytest
+
+import pandas as pd
+from pandas import DataFrame
+import pandas._testing as tm
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge_anti.py:6]: from pandas import testing as tm

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, some comments. needs some tests with other dtypes (datetime, datetime w/timezone, categorical and other EAs), doesnt' have to be on all parameterization a separate test is ok.

update docs where needed (doc-string & in merge.rst)

File: pandas/tests/reshape/merge/test_merge_anti.py (hunk 8 above)
Comment (debnathshoham) [reply pandas/tests/reshape/merge/test_merge_anti.py:9]: Actually pre-commit is not allowing that.
Screenshot 2021-08-19 at 6 54 57 PM
It also looks like this in other files
https://github.com/pandas-dev/pandas/blob/3af1a4fa27a7d48d8e6aea66bf2da6fdf2e1550e/pandas/tests/reshape/merge/test_merge.py#L34

File: pandas/core/reshape/merge.py (hunk 7 above)
Comment (debnathshoham) [reply pandas/core/reshape/merge.py:749]: not sure if I understand what you mean by, being a method on _MergeOperation.

**Comment by debnathshoham** [comment]
This is rendering little odd. I had to split it into two lines due to pep8 length limitation, but not sure how to ensure the render is in a single line.

```
how : {'left', 'right', 'outer', 'inner', 'cross',
  'anti_left', 'anti_right', 'anti_full'}, default 'inner'
```

Screenshot 2021-08-19 at 7 44 18 PM

File: pandas/tests/reshape/merge/test_merge_anti.py [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+                    index=["c", "d"],
+                ),
+            ),
+        ],
+    )
+    def test_basic_anti_index(self, how, expected):
+        left = DataFrame({"A": [1, 2, 3], "C": [5, 6, 7]}, index=["a", "b", "c"])
+        right = DataFrame({"B": [1, 2, 4], "C": [7, 8, 9]}, index=["a", "b", "d"])
+        result = merge(left, right, how=how, left_index=True, right_index=True)
+        tm.assert_frame_equal(result, expected)
```
Comment (attack68) [inline pandas/tests/reshape/merge/test_merge_anti.py:43]:
```suggestion
    @pytest.mark.parametrize(
        "how, exp_index, exp_values", [
            ("anti_left", ["c"], [3, 7, np.nan, np.nan]),
            ("anti_right", ["d"], [np.nan, np.nan, 4, 9]),
            ("anti_full", ["c", "d"], [[[3, 7, np.nan, np.nan], [np.nan, np.nan, 4, 9])
    ])
    def test_basic_anti_index(self, how, exp_index, exp_values):
        left = DataFrame({"A": [1, 2, 3], "C": [5, 6, 7]}, index=["a", "b", "c"])
        right = DataFrame({"B": [1, 2, 4], "C": [7, 8, 9]}, index=["a", "b", "d"])
        expected = DataFrame(exp_values, index=exp_index, columns=["A", "C_x", "B", "C_y"])
        result = merge(left, right, how=how, left_index=True, right_index=True)
        tm.assert_frame_equal(result, expected)
```

**Comment by debnathshoham** [comment]
hi guys, just checking if anyone got a chance to look into this..

**Review by attack68 — CHANGES_REQUESTED** [review]
The actual mechanics of this and the coding structure **look good to me** - I think it has been through a few iterations and is in a clean state now.

Since this is fairly complex logic, i.e. you are taking the anti-join input and manipulating it into a structure that can then process the join as usual having been pre-processed, I would, personally, add better comments and documentation, explaining the functions and different cases.

Additionally, the tests, even though they look a little unwieldy, are thorough, but there are no comments here either and I think they would be quite helpful just indicating the purpose of the test.

I would also add a sentence describing the examples you gave in the documentation.

**Comment by debnathshoham** [comment]
thanks @attack68 for the detailed review.
I have added a couple of more comments, would be happy to add more, if you can point somewhere you think is still unclear).

**Comment by debnathshoham** [comment]
I have added a couple of comments in the tests too. In the rest, I think the name of the tests are self-descriptive.

File: pandas/tests/reshape/merge/test_merge_anti.py [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+                    }
+                ).astype({"B_x": object}),
+                "anti_right",
+                None,
+                ["A"],
+                ["C"],
+            ),
+        ],
+    )
+    def test_anti_with_nan(self, expected, how, on, left_on, right_on):
```
Comment (attack68) [inline pandas/tests/reshape/merge/test_merge_anti.py:304]: what is this test doing that the other tests above do not capture?

File: pandas/tests/reshape/merge/test_merge_anti.py [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+                        "B_y": [np.nan, np.nan, pd.NA, 2],
+                    }
+                ).astype({"B_x": object}),
+                "anti_full",
+                "A",
+                "C",
+            ),
+        ],
+    )
+    def test_anti_with_nan_and_NA(self, expected, how, left_on, right_on):
```
Comment (attack68) [inline pandas/tests/reshape/merge/test_merge_anti.py:385]: Is this test purely testing `pd.NA`? 

In this case it seems overkill. Logically speaking if the tests above imply everything is working for `nan`, and a simple test shows `pd.Na` works equivalently to `nan` then all tests above should work for `pd.NA` so we don't have to repeat everything again.

File: pandas/tests/reshape/merge/test_merge_anti.py [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+                        "vals_y": [np.nan, np.nan, 17, 21],
+                    },
+                    index=pd.date_range("1/7/2010", periods=2, freq="2d").union(
+                        pd.date_range("1/2/2010", periods=2, freq="2d")
+                    ),
+                ),
+            ),
+        ],
+    )
+    def test_anti_datetime(self, how, expected):
```
Comment (attack68) [inline pandas/tests/reshape/merge/test_merge_anti.py:422]: same for datetime. It feels like you are testing everything everytime.
If you are confident from previous tests that the different "anti_xxx" work is there a sufficient reason to suggest that two might work and the other one not for a different dtype? 

If not then only one test is needed, not 3, and you might be able to combine with datetime_tz and make this much more readable.

File: pandas/core/reshape/merge.py [hunk 13]
```diff
@@ -743,6 +745,46 @@ def get_result(self) -> DataFrame:
...
+        """
+        if self.left_index and self.right_index:
+            # Merge using `right_index` and `left_index`
+            join_index_l, join_index_r, self.how = _anti_helper(
+                self.left.index, self.right.index, self.how
+            )
+        elif self.on is not None or (
+            None not in self.left_on and None not in self.right_on
+        ):
+            # Merge using `on` or `left_on` and `right_on`
```
Comment (attack68) [inline pandas/core/reshape/merge.py:762]: to be honest these comments are obvious, as you said. it was more the general concept that you could have expanded in the main docstring, why you have implemented it this way and what is the advantage. Just in case you come back to look at it after two years, or another developer comes in and tries to figure out what you are doing.

**Review by attack68 — CHANGES_REQUESTED** [review]
Im pretty happy with this as is, but I do think it can be streamlined.

File: pandas/core/reshape/merge.py (hunk 7 above)
Comment (jreback) [reply pandas/core/reshape/merge.py:748]: this needs to execute in the `get_result()` portion, NOT during validation. validation shouldn't dot he actual computation.

IOW during _get_join_info could dispatch based on `how`.

**Review by jreback — CHANGES_REQUESTED** [review]
this needs to follow the existing workflow. as-is its very hard to follow what you are trying to do. the pattern is validate the merge, *then* actually do the operation. you have mixed the two.

File: pandas/core/reshape/merge.py (hunk 7 above)
Comment (debnathshoham) [reply pandas/core/reshape/merge.py:748]: the actual calculation is still happening in `get_result()`. 
I have just changed the configuration of `left`, `right` and `how` to utilise the already existing joining methods. E.g. For `anti_left` I have changed the `left`, `right` and `how` to `left` such that `left join` would give the result

File: folder/subfolder/out.csv [hunk 14]
```diff
@@ -0,0 +1,3 @@
+,name,mask,weapon
```
Comment (attack68) [inline folder/subfolder/out.csv:1]: why have we created a csv file?

Comment (debnathshoham) [reply folder/subfolder/out.csv:1]: ah, this was by mistake.. will remove this

File: pandas/tests/reshape/merge/test_merge_anti.py (hunk 11 above)
Comment (debnathshoham) [reply pandas/tests/reshape/merge/test_merge_anti.py:385]: actually this was to check `np.nan` is not matched with `pd.NA`

File: pandas/tests/reshape/merge/test_merge_anti.py (hunk 12 above)
Comment (debnathshoham) [reply pandas/tests/reshape/merge/test_merge_anti.py:422]: sure, I can test only one, and remove the other two.
I had separated the one with tz for readability (can combine into one if you think otherwise).

File: pandas/core/reshape/merge.py [hunk 15]
```diff
@@ -1450,6 +1492,59 @@ def _validate(self, validate: str) -> None:
...
+    if not isinstance(_left, Index):
+        if len(_left.columns) == 1:
+            _left = Index(_left.values.flatten(), dtype=_left.dtypes[0])
+        else:
+            _left = MultiIndex.from_frame(_left)
+    if not isinstance(_right, Index):
+        if len(_right.columns) == 1:
+            _right = Index(_right.values.flatten(), dtype=_right.dtypes[0])
+        else:
+            _right = MultiIndex.from_frame(_right)
```
[the remark this answers is not in the document]
Comment (debnathshoham) [reply pandas/core/reshape/merge.py:1534]: reverted this commit. 
original `_left` and `_right` were remaining unchanged, so it was throwing error on L1543-1544

File: pandas/tests/reshape/merge/test_merge_anti.py (hunk 10 above)
Comment (debnathshoham) [reply pandas/tests/reshape/merge/test_merge_anti.py:304]: this was for mixed dtypes.
(added couple of more comments for these)

File: pandas/core/reshape/merge.py (hunk 13 above)
Comment (debnathshoham) [reply pandas/core/reshape/merge.py:762]: the comments in merge.py looks sufficient to me. But again, I wrote it and a second pair of eyes can give a fresh perspective.
Do you think the logic in merge.py is too difficult to follow? (if you could point to any specific portion)

File: pandas/tests/reshape/merge/test_merge_anti.py (hunk 12 above)
Comment (attack68) [reply pandas/tests/reshape/merge/test_merge_anti.py:424]: my point being if you consider all the possible aspects to test:

dtypes: [float, int, bool, datetime, ea, categorical, etc.]
how: [anti_left, anti_right, anti_full]
kwargs: [on, left_on, right_on, left_index, right_index]
multiindex: [yes, no]
nan: [nan, pd.Na, None]

There is approximately over 500 different parametrised test combinations here. You seem to be doing around 50. I would try to get this down to about 25 or less, with structured logic.

For example, if you test that "how" works as expected for "anti_left", "anti_right" and "anti_full", and separately you show that "anti_full" works for different dtypes, you can make the logical assumption that for those additional dtypes "anti_left" and "anti_right" also work, so you dont have to test them.

I think it would be enough to show that pd.Na and nan and None are not treated as equal in a single test, and this would be enough to assume it works for all cases.

Comment (debnathshoham) [reply pandas/tests/reshape/merge/test_merge_anti.py:424]: I beg to differ here.
True that I have covered a small part of the complete set of combinations, but I don't agree that reducing that is a good idea. Different paths are taken at different places depending upon every param.
IOW just because something worked with single dtype wouldn't ensure it worked with mixed / EA

Comment (attack68) [reply pandas/tests/reshape/merge/test_merge_anti.py:424]: OK, we probably share different philosophies. Mine is that pandas is a large package and if all possible developments were tested thoroughly with all params the tests might take a very long time. Ultimately the tests are there to prevent future regressions, they are not in place to test that everything works as developed: you can do that independently. Whilst possible, I would view it unlikely that someone introduces a regression which only breaks "anti_right" and only for one particular dtype. I think more likely it either breaks "anti_right" for all dtypes, or all the "anti_xx" combinations break for a single dtype. That is my logic, and why I would target those fewer tests.

But I acknowledge it is subjective and won't push it. :)

File: pandas/core/reshape/merge.py (hunk 7 above)
Comment (debnathshoham) [reply pandas/core/reshape/merge.py:747]: hi @jreback - wondering if you got a chance to look at this? and if this implementation is fine?

Comment (jreback) [reply pandas/core/reshape/merge.py:747]: no the implementation is quite convoluted 

i am not sure how to fix what u have here

**Comment by jreback** [comment]
conceptually ok with adding this. but the implementation needs tight integration with the existing. you are bolting it on in an unmaintainable manner. if you want to continue, merge master and update.
